app/src/main/java/org/mozilla/rocket/content/news/ui/NewsFragment.kt

```diff
@@ -77,40 +78,60 @@ class NewsFragment : Fragment() {
     }
 
     private fun initNewsList() {
-        newsAdapter = DelegateAdapter(
+        newsAdapter = object : DelegatePagedListAdapter(
             AdapterDelegatesManager().apply {
                 add(NewsSourceLogoUiModel::class, R.layout.item_news_source_logo, NewsSourceLogoAdapterDelegate())
                 add(NewsUiModel::class, R.layout.item_news, NewsAdapterDelegate(getCategory(), newsViewModel))
-            }
-        )
-
-        news_list.apply {
-            adapter = newsAdapter
-        }
+            },
+            object : DiffUtil.ItemCallback<DelegateAdapter.UiModel>() {
+                override fun areItemsTheSame(oldItem: DelegateAdapter.UiModel, newItem: DelegateAdapter.UiModel): Boolean {
+                    if (oldItem::class != newItem::class) {
+                        return false
+                    }
+                    return when (oldItem) {
+                        is NewsSourceLogoUiModel -> true
+                        is NewsUiModel -> {
+                            newItem as NewsUiModel
+                            oldItem.componentId == newItem.componentId
+                        }
+                        else -> error("unexpected type: ${oldItem.javaClass}")
+                    }
+                }
 
-        (news_list.layoutManager as LinearLayoutManager).let {
-            news_list.addOnScrollListener(object : RecyclerView.OnScrollListener() {
-                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
-                    super.onScrolled(recyclerView, dx, dy)
-                    val totalItemCount = it.itemCount
-                    val visibleItemCount = it.childCount
-                    val lastVisibleItem = it.findLastVisibleItemPosition()
-                    if (visibleItemCount + lastVisibleItem + NEWS_THRESHOLD >= totalItemCount) {
-                        loadMore()
+                override fun areContentsTheSame(oldItem: DelegateAdapter.UiModel, newItem: DelegateAdapter.UiModel): Boolean {
+                    return when (oldItem) {
+                        is NewsSourceLogoUiModel -> true
+                        is NewsUiModel -> {
+                            newItem as NewsUiModel == oldItem
+                        }
+                        else -> error("unexpected type: ${oldItem.javaClass}")
                     }
                 }
-            })
+            }
+        ) {
+            override fun getItemId(position: Int): Long {
+                return when (val item = getItem(position)) {
+                    is NewsSourceLogoUiModel -> item.hashCode().toLong()
+                    is NewsUiModel -> item.componentId.hashCode().toLong()
+                    null -> super.getItemId(position)
+                    else -> error("unexpected type: ${item.javaClass}")
+                }
+            }
+        }.apply {
+            setHasStableIds(true)
+        }
 
-            news_list.monitorScrollImpression(telemetryViewModel)
```

Comment:
Please help to add the `monitorScrollImpression` back or we will lose the scroll impression telemetry.

Reply:
fixed, thanks!